Utilities/DataLoading/EdFi.LoadTools/ApiClient/ResourceToInterchangeMapProvider.cs

```diff
@@ -2,69 +2,213 @@
 // Licensed to the Ed-Fi Alliance under one or more agreements.
 // The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
 // See the LICENSE and NOTICES files in the project root for more information.
- 
+
+using System;
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
 using System.Linq;
+using System.Text.RegularExpressions;
+using System.Xml;
 using System.Xml.Schema;
 using EdFi.LoadTools.Engine;
+using log4net;
 
 namespace EdFi.LoadTools.ApiClient
 {
-    public class ResourceToInterchangeMapProvider
+    public interface IResourceToFileTypeInfoProvider
     {
-        private XmlSchemaSet _xmlSchemaSet;
+        Dictionary<string, List<FileTypeInfo>> GetResourceByFileTypeInfo();
+    }
 
-        private readonly Dictionary<string, List<InterchangeInfo>> _resourceToInterchangeDictionary = new Dictionary<string, List<InterchangeInfo>>();
-        private readonly IResourceStreamFactory _resourceStreamFactory;
+    public class ResourceToFileTypeInfoProvider : IResourceToFileTypeInfoProvider
+    {
+        private readonly IDataConfiguration _dataConfiguration;
+        private readonly Regex _interchangeNameRegex = new Regex(Constants.InterchangeNameRegex, RegexOptions.Compiled);
+        private readonly ILog _log = LogManager.GetLogger(nameof(ResourceToFileTypeInfoProvider));
+        private readonly Dictionary<string, List<FileTypeInfo>> _resourceByFileTypeInfo =
+            new Dictionary<string, List<FileTypeInfo>>();
+        private readonly Lazy<Dictionary<string, List<FileContext>>> _resourceNameByFileContext;
+        private readonly XmlSchemaSet _xmlSchemaSet;
+        private readonly IXsdConfiguration _xsdConfiguration;
 
-        public ResourceToInterchangeMapProvider(XmlSchemaSet xmlSchemaSet, IResourceStreamFactory resourceStreamFactory)
+        public ResourceToFileTypeInfoProvider(XmlSchemaSet xmlSchemaSet, IDataConfiguration dataConfiguration,
+            IXsdConfiguration xsdConfiguration)
         {
             _xmlSchemaSet = xmlSchemaSet;
-            _resourceStreamFactory = resourceStreamFactory;
+            _dataConfiguration = dataConfiguration;
+            _xsdConfiguration = xsdConfiguration;
+
+            _resourceNameByFileContext =
+                new Lazy<Dictionary<string, List<FileContext>>>(GetFileTypeByFileContext);
         }
 
-        public Dictionary<string, List<InterchangeInfo>> GetResourceToInterchangeMap()
+        public Dictionary<string, List<FileTypeInfo>> GetResourceByFileTypeInfo()
```

Comment:
I'm not understanding the need for the two layers.  We have this dictionary which gives a list of FileTypeInfo objects per resource, then each FileTypeInfo has a list of FileContexts.  Why do we need the extra layer of "FileType" (Interchange) at all?  I feel like we've got the relationships flip flopped here.  Why not just build and maintain an internal List of FileContexts, and as part of processing it also track which resources are actually in the file?  That way you only read the files that actually have instances of the resource inside it, and you have the list of resources a file contains as part of the FileContext.  Then provide a method on this provider to look up the list of file contexts to process for a given resource name.

Reply:
done, the methods returns only a list of filetypeinfos